[Feature] add retry for http client

machichima opened the pull request:
- Enable retry in `executeRequest` in `pkg/http/client.go`
- Add unit test to ensure the retry function works as expected

## Why are these changes needed?

Currently, the apiserver http client does not support retry. This PR enables the retry feature in http client.

## Related issue number

Part of #3344

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
  - [ ] Unit tests
  - [ ] Manual tests
  - [ ] This PR is not tested :(

machichima commented:
Hi @dentiny , PTAL
Thanks!

dentiny reviewed (changes requested):
My comments seems to be lost somewhere...

Before implementing the retry, I think there're a few things we need to confirm:
- What to retry and what not to retry? For example, permanent errors (i.e. permission denied) shouldn't be retried.
- How to retry? Apart from the max retry you added, we also need set overall timeout, exponential backoff, etc.

dentiny commented:
let me know if you want a discussion, or you feel the PR is ready to review :)

machichima commented:
Hi @dentiny ,
I've modified to retry only on specific http status errors and added backoff.

For the overall timeout, I think it is now set when specifying `http.Client` before sending to `NewKuberayAPIServerClient`, do we also need to take care of that in `NewKuberayAPIServerClient`?

https://github.com/ray-project/kuberay/blob/5f593bc20b8d7f58f2f7b92d38bcef4b568ec321/apiserver/test/e2e/types.go#L95

dentiny commented on apiserver/pkg/http/client.go:
```diff
@@ -27,6 +29,10 @@ type KuberayAPIServerClient struct {
 	// Store http request handling function for unit test purpose.
 	executeHttpRequest func(httpRequest *http.Request, URL string) ([]byte, *rpcStatus.Status, error)
 	baseURL            string
+	maxRetry           int
```
curious do we control overall timeout?

machichima replied:
Hi @dentiny 

The overall timeout is set in the `httpClient` before sending to `kuberayHTTP.NewKuberayAPIServerClient` now.

https://github.com/ray-project/kuberay/blob/5f593bc20b8d7f58f2f7b92d38bcef4b568ec321/apiserver/test/e2e/types.go#L95-L97

Do you think we should set the `maxTimeout` in `KuberayAPIServerClient` and override user's config if the timeout set in `httpClient` is larger than this value?

dentiny replied:
I'm not sure that's "overall timeout" for all retry attempts?
Sounds like timeout for internal http client operation timeout

dentiny replied:
btw do you mind if we making retry related fields a separate config struct?

machichima replied:
> I'm not sure that's "overall timeout" for all retry attempts?
Sounds like timeout for internal http client operation timeout

Ah yes, it is indeed the timeout for one http request. I will add overall timeout then

machichima replied:
> btw do you mind if we making retry related fields a separate config struct?

Sure!

dentiny commented on apiserver/pkg/http/client.go:
```diff
@@ -27,6 +29,10 @@ type KuberayAPIServerClient struct {
 	executeHttpRequest func(httpRequest *http.Request, URL string) ([]byte, *rpcStatus.Status, error)
 	baseURL            string
+	maxRetry           int
+	backoffBase        float64
```
```suggestion
	backoffFactor        float64
```

dentiny commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +648,78 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+
+		defer func() {
+			if closeErr := response.Body.Close(); closeErr != nil {
+				klog.Errorf("Failed to close http response body because %+v", closeErr)
```
nit: I understand it's not your issue, but let's keep error message consistent and use either `Failed` or `failed`.

dentiny commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +648,78 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+
+	for attempt := 0; attempt < krc.maxRetry; attempt++ {
+		response, err := krc.httpClient.Do(httpRequest)
+		// do not retry for Go Error
```
Could you please explain what do you mean by "Go Error"?
According to the [doc](https://pkg.go.dev/net/http#Client.Do): 
```
An error is returned if caused by client policy (such as CheckRedirect), or failure to speak HTTP (such as a network 
connectivity problem). A non-2xx status code doesn't cause an error.
```

dentiny commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +648,78 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+		}()
+
+		bodyBytes, err := io.ReadAll(response.Body)
+		// do not retry for Go Error
```
same here

dentiny commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +648,78 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+	var lastErr error
+	var lastStatus *rpcStatus.Status
+
+	for attempt := 0; attempt < krc.maxRetry; attempt++ {
```
let's add a comment on when to retry and when not, for your current implementation, only http error status is retried

dentiny commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +648,78 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
-		return nil, status, &KuberayAPIServerClientError{
-			HTTPStatusCode: response.StatusCode,
+
+		if response.StatusCode != http.StatusOK {
```
let's do early return to reduce identation.

```go
if statusCode == OK {
  return bytes
}
// error handling
```

dentiny commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +648,78 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+
+		if response.StatusCode != http.StatusOK {
+			status, err := krc.extractStatus(bodyBytes)
+			// do not retry for Go Error
```
same here

dentiny commented on apiserver/pkg/http/client_test.go:
```diff
@@ -1,17 +1,65 @@
+		BackoffFactor:  util.HTTPClientDefaultBackoffBase,
+		InitBackoff:    util.HTTPClientDefaultInitBackoff,
+		MaxBackoff:     util.HTTPClientDefaultMaxBackoff,
+		OverallTimeout: util.HTTPClientOverallTimeout,
```
also name it as "default"?

dentiny commented on apiserver/pkg/http/client_test.go:
```diff
@@ -59,3 +115,188 @@ func TestAPIServerClientError(t *testing.T) {
+			expectBody: nil,
+		},
+		{
+			name:     "Stops on Go error",
```
nit: rename

dentiny commented on apiserver/pkg/util/config.go:
```diff
@@ -24,6 +24,17 @@ const (
 	RayClusterDefaultImageRepository = "rayproject/ray"
+
+	// Max retry times for HTTP Client
+	HTTPClientDefaultMaxRetry = 3
```
just FYI (no change requested), the default overall and max timeout reads like your max retry is more proper to be 4.

dentiny commented on apiserver/pkg/http/client_test.go:
```diff
@@ -1,17 +1,65 @@
+	returnDoError  bool
+}
+
+func (m *mockTransport) RoundTrip(_ *http.Request) (*http.Response, error) {
```
nit: add a documentation on behavior, especially when `m.statusSequence` all consumed, we return OK status.

dentiny commented on apiserver/pkg/http/client_test.go:
```diff
@@ -1,17 +1,65 @@
+		respBody = statusMsg
+	}
+
+	m.callCount++
```
nit: do you mind put increment right after L33? I'm worried we might early return in the middle in the future.

dentiny reviewed (approved):
Thanks for the thorough testing!
(no need to be in this PR) On retry, I generally suggest to make it a separate util function.
Apart from the retry config and context with timeout, it should also take a functor to decide whether retriable or not.

In detail
```go
func RunWithRetry(f func() error, isRetriable func(err error) bool, cfg RetryConfig);
```

dentiny commented:
> Hi @dentiny
> 
> I found that my original retry logic is incorrect. The max attempt times should be maxRetry + 1. Originally, I use `<` in the following for loop, so the attempt time is `maxRetry`. I modified as follow:
> 
> https://github.com/ray-project/kuberay/blob/7e0d6ad77528843ae0921cc193fbd1781c18b219/apiserver/pkg/http/client.go#L664

Thanks! Then can we add a unit test with max retry count 0, to make sure functor does execute?

machichima commented:
@dentiny I modified the `TestAPIServerClientRetry` to let it validate the `callCount` is `expectCallCount` and add the test case for `maxRetry` = 0

Thanks!

dentiny reviewed (approved):
Thanks!

rueian commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +650,90 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
-	}
-	if response.StatusCode != http.StatusOK {
+
+		defer func() {
+			if closeErr := response.Body.Close(); closeErr != nil {
+				klog.Errorf("failed to close http response body because %+v", closeErr)
+			}
+		}()
```
We can't use defer to close the per loop response.Body in this way. deferred blocks are only be executed at the end of `executeRequest`. Consider using this instead:

```go
doReq = func() (response, error) {
    ...
    defer func() {
        response.Body.Close()
    }()
    ....
}
for attempt := 0; attempt <= krc.retryCfg.MaxRetry; attempt++ {
    response, err := doReq(httpRequest)
}
```

machichima commented:
cc @kevin85421 for final review

The failing CI for autoscaler seems to be flaky test, while this PR only modify apiserver part

kevin85421 reviewed (approved):
stamp

kevin85421 commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +649,92 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+	switch statusCode {
+	case http.StatusRequestTimeout, // 408
+		http.StatusTooManyRequests,    // 429
+		http.StatusBadGateway,         // 502
```
500?

machichima replied:
Added!

kevin85421 commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +649,92 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+}
+
+func isRetryableHTTPStatus(statusCode int) bool {
+	switch statusCode {
```
How about creating a `map[T]struct{}` for all retryable HTTP status codes and looking up the map? The `switch` statement and the utility function are not necessary in my opinion.

dentiny replied:
Or even use an array.
Array outperforms hash map when item number is small (no need to calculate hash value, better cache locality)

kevin85421 commented on apiserver/pkg/http/client.go:
```diff
@@ -638,29 +649,92 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+
+	// Only retry for HTTP status codes defined as retryable in isRetryableHTTPStatus().
+	for attempt := 0; attempt <= krc.retryCfg.MaxRetry; attempt++ {
+		bodyBytes, statusCode, err := doReq()
```
Is this a blocking operation? If so, could it block here indefinitely in some cases, even if OverallTimeout is exceeded? I don't know whether `doReq` will also exit when `cancel` is called.

machichima replied:
I have a test here for overall timeout. I think `doReq` will indeed exist when timeout

https://github.com/ray-project/kuberay/blob/617c0c894f09bd17e6fac5626b09ee0286e1635f/apiserver/pkg/http/client_test.go#L292

kevin85421 replied:
@dentiny @rueian do we have timeout for `doReq`?

rueian replied:
yes, we have the timeout here https://github.com/ray-project/kuberay/pull/3551/files#diff-161b63aa5758f415dad6d6277186bbb9609ccb899b8e32efca3f9c452aae7810R662-R665

machichima commented on apiserver/pkg/http/client.go:
```diff
@@ -637,30 +648,89 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
+	var lastStatus *rpcStatus.Status
+
+	doReq := func() ([]byte, int, error) {
+		response, err := krc.httpClient.Do(httpRequest)
```
Fixed! Thanks!

rueian reviewed (approved):
LGTM!

kevin85421 reviewed (approved):
@machichima can you open an issue to track the progress of updating documents? Then, I will merge this PR.

machichima commented:
@kevin85421 Thank you! Just opened a doc issue here: https://github.com/ray-project/kuberay/issues/3883
